rust/processor/src/db_writer.rs

```diff
@@ -0,0 +1,325 @@
+            .with_label_values(&[processor_name, table_name])
+            .set(insertion_timer.elapsed().as_secs_f64());
+
+        DB_INSERTION_ROWS_COUNT
+            .with_label_values(&[processor_name, table_name])
+            .inc_by(self.len() as u64);
+        res
+    }
+}
+
+pub fn diesel_error_to_metric_str(error: &Error) -> &'static str {
+    match error {
+        Error::DatabaseError(kind, _) => match kind {
+            DatabaseErrorKind::UniqueViolation => "DatabaseError::UniqueViolation",
+            DatabaseErrorKind::ForeignKeyViolation => "DatabaseError::ForeignKeyViolation",
+            DatabaseErrorKind::UnableToSendCommand => "DatabaseError::UnableToSendCommand",
+            DatabaseErrorKind::SerializationFailure => "DatabaseError::SerializationFailure",
+            DatabaseErrorKind::ReadOnlyTransaction => "DatabaseError::ReadOnlyTransaction",
+            DatabaseErrorKind::NotNullViolation => "DatabaseError::NotNullViolation",
+            DatabaseErrorKind::CheckViolation => "DatabaseError::CheckViolation",
+            DatabaseErrorKind::ClosedConnection => "DatabaseError::ClosedConnection",
+            DatabaseErrorKind::Unknown => "DatabaseError::Unknown",
+            _ => "DatabaseError::Other",
+        },
+        Error::NotFound => "NotFound",
+        Error::QueryBuilderError(_) => "QueryBuilderError",
+        Error::DeserializationError(_) => "DeserializationError",
+        Error::SerializationError(_) => "SerializationError",
+        Error::InvalidCString(_) => "InvalidCString",
+        Error::RollbackTransaction => "RollbackTransaction",
+        Error::AlreadyInTransaction => "AlreadyInTransaction",
+        Error::RollbackErrorOnCommit { .. } => "RollbackErrorOnCommit",
+        Error::NotInTransaction => "NotInTransaction",
+        Error::BrokenTransactionManager => "BrokenTransactionManager",
+        _ => "Other",
+    }
+}
+
+// This is a fake table for default fallback chunk size
+pub struct FakeTable {}
+impl field_count::FieldCount for FakeTable {
+    fn field_count() -> usize {
+        100
+    }
+}
+
+// A holder struct for processors db writing so we don't need to keep adding new params
+#[derive(Clone)]
+pub struct DbWriter {
+    pub processor_name: &'static str,
+    pub db_pool: PgDbPool,
+    pub query_sender: AsyncSender<QueryGenerator>,
+    pub per_table_chunk_sizes: AHashMap<String, usize>,
+    pub skip_tables: AHashSet<String>,
+}
+
+impl DbWriter {
+    pub fn new(
+        processor_name: &'static str,
+        db_pool: PgDbPool,
+        query_sender: AsyncSender<QueryGenerator>,
+        per_table_chunk_sizes: AHashMap<String, usize>,
+        skip_tables: AHashSet<String>,
+    ) -> Self {
+        Self {
+            processor_name,
+            db_pool,
+            query_sender,
+            per_table_chunk_sizes,
+            skip_tables,
+        }
+    }
+
+    pub fn chunk_size<Item: field_count::FieldCount>(&self, table_name: &str) -> usize {
```

Comment:
not related to this PR, but one day we probably also need to have limit on bytes size in addition to # of rows.